s3proxy/internal/router/router.go

```diff
@@ -0,0 +1,459 @@
+/*
+Copyright (c) Edgeless Systems GmbH
+
+SPDX-License-Identifier: AGPL-3.0-only
+*/
+
+/*
+Package router implements the main interception logic of s3proxy.
+It decides which packages to forward and which to intercept.
+
+The routing logic in this file is taken from this blog post: https://benhoyt.com/writings/go-routing/#regex-switch.
+We should be able to replace this once this is part of the stdlib: https://github.com/golang/go/issues/61410.
+*/
+package router
+
+import (
+	"bytes"
+	"crypto/md5"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/xml"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/edgelesssys/constellation/v2/internal/logger"
+	"github.com/edgelesssys/constellation/v2/s3proxy/internal/s3"
+)
+
+var (
+	regexen = make(map[string]*regexp.Regexp)
+	relock  sync.Mutex
+)
+
+// Router implements the interception logic for the s3proxy.
+type Router struct {
+	region string
+	log    *logger.Logger
+}
+
+// New creates a new Router.
+func New(region string, log *logger.Logger) Router {
+	return Router{region: region, log: log}
+}
+
+// Serve implements the routing logic for the s3 proxy.
+// It intercepts GetObject and PutObject requests, encrypting/decrypting their bodies if necessary.
+// All other requests are forwarded to the S3 API.
+// Ideally we could separate routing logic, request handling and s3 interactions.
+// Currently routing logic and request handling are integrated.
+func (r Router) Serve(w http.ResponseWriter, req *http.Request) {
+	var h http.Handler
+	var key string
+	var bucket string
+
+	client, err := s3.NewClient(r.region)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	path := req.URL.Path
+	switch {
```

Comment:
I'd refactor this switch case:
1. factor out the function calls like `containsBucket(req.Host)` to before the switch/case and instead assign those to bool variables which have a clear name. This should improve readability.
2. refactor each case "body" to its own function like `handle...()`.

Reply:
Code should a lot more concise now. Sorry for the convoluted initial version :see_no_evil: